_posts/2016-12-27-angular-2-advance-testing-with-custom-matchers.md

```diff
@@ -0,0 +1,650 @@
+
+{% highlight js %}
+{% raw %}
+expectNativeEl(fixture).toHaveCssStyle({
+  'display': 'flex',
+  'flex-direction': 'row',
+  'box-sizing': 'border-box'
+});
+{% endraw %}
+{% endhighlight %}
+
+Creating a matcher in JavaScript is documented on the Jasmine site. Our challenge is the harder 
+goal of creating a *custom matcher* implemented in TypeSript and well-defined types.
+
+First I need to enhance the `expect()` API and then implement custom matchers.
+
+The global Jasmine `expect()` method normally returns `<any>` value. To use custom matchers 
+(with types), I want the **expect** to support returning a custom matcher.
+
+Here is how I do that in a `custom-matchers.ts` module:
+
+{% highlight js %}
+{% raw %}
+declare var global: any;
+const _global = <any>(typeof window === 'undefined' ? global : window);
+
+/**
+ * Extend the API to support chaining with custom matchers
+ */
+export const expect: (actual: any) => NgMatchers = <any> _global.expect;
+
+/**
+ * Jasmine matchers that support checking custom CSS conditions.
+ */
+export interface NgMatchers extends jasmine.Matchers {
+  toHaveCssStyle(expected: {[k: string]: string}|string): boolean;
+  not: NgMatchers;
+}
+
+/**
+ * Implementation of 1...n custom matchers
+ */
+export const customMatchers: jasmine.CustomMatcherFactories = {
+
+  toHaveCssStyle: function () {
+    return {
+      compare: function (actual: any, styles: {[k: string]: string}|string) {
+        let allPassed: boolean;
+        if (typeof styles === 'string') {
+          allPassed = getDOM().hasStyle(actual, styles);
+        } else {
+          allPassed = Object.keys(styles).length !== 0;
+          Object.keys(styles).forEach(prop => {
+            allPassed = allPassed && getDOM().hasStyle(actual, prop, styles[prop]);
+          });
+        }
+
+        return {
+          pass: allPassed,
+          get message() {
+            const expectedValueStr = typeof styles === 'string' ? styles : JSON.stringify(styles);
+            return `Expected ${actual.outerHTML} ${!allPassed ? ' ' : 'not '}to contain the
+                      CSS ${typeof styles === 'string' ? 'property' : 'styles'} "${expectedValueStr}"`;
+          }
+        };
+      }
+    };
+  }
+
+};
+{% endraw %}
+{% endhighlight %}
+
+With the above definitions, I can now use `expect(...).toHaveCssStyles(...)` without any 
+TypeScript complaints.
+
+Wait... I need one more addition to our custom-matcher code. Notice the call to 
+`getDOM().hasStyle()` ?
+
+Where does `getDOM` come from ? After some inspection of the @angular/core code, I was able to 
+get a reference to the special Angular DOM Adapter:
+
+{% highlight js %}
+{% raw %}
+import {__platform_browser_private__} from '@angular/platform-browser';
+const getDOM = __platform_browser_private__.getDOM;
+{% endraw %}
+{% endhighlight %}
+
+Now I am *golden* with features. Let's import and use our jasmine matcher.
```

Comment:
We should mention here that this code relies on private APIs and can break in the future.